man/dbls.1
.TH DBLS 1 "12 September 1995"
.FB
.SH NAME
dbls \- list all \fIFirstBase\fP databases and associated files in a directory
.SH SYNOPSIS
.B dbls
[ \fIdirectory\fP ... ]
.SH DESCRIPTION
.I Dbls
is a command that lists all \fIFirstBase\fP databases and 
associated files in a directory. The default directory displayed is the
current working directory.
.PP
Listed files include databases, database dictionaries, database maps, 
indexes, index dictionaries, index generator dictionaries, printout
dictionaries, screen dictionaries and view dictionaries.
.SH FILES
.PD 0
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary.
.TP 10
*.map
\fIFirstBase\fP database record map.
.TP 10
*.idx
\fIFirstBase\fP index.
.TP 10
*.idict
\fIFirstBase\fP index dictionary.
.TP 10
*.idicti
\fIFirstBase\fP index generator dictionary.
.TP 10
*.idictp
\fIFirstBase\fP printout dictionary.
.TP 10
*.idictl
\fIFirstBase\fP label dictionary.
.TP 10
*.idictc
\fIFirstBase\fP conversion dictionary.
.TP 10
*.sdictc
\fIFirstBase\fP screen dictionary.
.TP 10
*.vdict
\fIFirstBase\fP view dictionary.
.TP 10
*.noadd
add record lock.
.TP 10
*.nodel
delete record lock.
.PD
.SH SEE ALSO
firstbase(5), dictionaries(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbls.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbls.c,v 9.1 2001/01/16 02:46:47 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char *Dbls_sid = "%W% %G% FB";
#endif

#include <fb.h>
#include <fb_vars.h>
                                                                  
#if HAVE_DIRENT_H
# include <dirent.h>
# define NAMLEN(dirent) strlen((dirent)->d_name)
#else
# define dirent direct
# define NAMLEN(dirent) (dirent)->d_namlen
# if HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif
# if HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif
# if HAVE_NDIR_H
#  include <ndir.h>
# endif
#endif


typedef struct dirent dstruct;

dstruct *pp;

static void onels();
static display();
static dload();
static dshow();
static compar();
static void afile();
static void install();

static char *DDOT = ".";

#define NTYPES		19		/* MUST BE MULTIPLE OF 3 + 1 */
#define MAXFILES	500

static char *UTYPES[NTYPES] = {		/* MUST BE MULTIPLE OF 3 + 1 */
   0,
   "========",				/* 1 dbase */
   "===================",		/* 2 ddict */
   "============",			/* 3 map */
   "=====",				/* 4 idx */
   "================",			/* 5 idict */
   "==========================",	/* 6 idicti */
   "===============",			/* 7 view */
   "=================",			/* 8 screen */
   "=====================",		/* 9 conversion */
   "========",				/* 10 printout */
   "===================",		/* 11 print dict */
   "=========",				/* 12 emit */
   "======",				/* 13 labels */
   "================",			/* 14 label dict */
   "=================",			/* 15 update */
   "==============",			/* 16 */
   "===========",			/* 17 */
   0
   };

static char *FTYPES[NTYPES] = {		/* MUST BE MULTIPLE OF 3 + 1 */
   0,
   "Database",				/* 1 */
   "Database Dictionary",		/* 2 */
   "Database Map",			/* 3 */
   "Index",				/* 4 */
   "Index Dictionary",			/* 5 */
   "Index Generator Dictionary",	/* 6 */
   "View Dictionary",			/* 7 */
   "Screen Dictionary",			/* 8 */
   "Conversion Dictionary",		/* 9 */
   "Printout",				/* 10 */
   "Printout Dictionary",		/* 11 */
   "Emit File",				/* 12 */
   "Labels",				/* 13 */
   "Label Dictionary",			/* 14 */
   "Update Dictionary",			/* 15 */
   "No-Delete Lock",			/* 16 */
   "No-Add Lock",			/* 17 */
   0
   };

struct elem {
   char *fname;
   struct elem *next;
   struct elem *prev;
   };

struct elem *e_files[NTYPES] = {NULL};

/*
 * dbls - 
 *	this version expects 4.2 directory routines opendir, readdir, etc.
 *	these routines are available in a library usually called libndir.a
 */

   main(argc, argv)
      int argc;
      char **argv;
   
      {
	 int i;

         fb_setup();
	 for (i = 0; i < NTYPES; i++)
	    e_files[i] = NULL;
         if (argc <= 1)
            onels(DDOT);
         else
            for (--argc, ++argv; argc > 0; argc--){
               onels(*argv);
               argv++;
               }
      }

/*
 * onels - do one dbls command, on DDOT or given argument (directory)
 */

   static void onels(s)
      char *s;

      {
	 DIR *dirpp;
	 dstruct *pp;

	 dirpp = opendir(s);
         if (dirpp == NULL){
            fb_serror(FB_CANT_OPEN, s, NIL);
            return;
            }
         pp = readdir(dirpp);
	 for (; pp != NULL;  pp = readdir(dirpp))
	    if (pp->d_name[0] != '.')
	       afile(pp->d_name);
	 closedir(dirpp);
	 display();
      }

/* display - driver to display the argvecs three at a time */
   static display()
   
      {
         int n, typ;
	 char **p1, **p2, **p3, **argvec1, **argvec2, **argvec3;
	 
	 argvec1 = (char **) fb_malloc(sizeof(char *) * MAXFILES);
	 argvec2 = (char **) fb_malloc(sizeof(char *) * MAXFILES);
	 argvec3 = (char **) fb_malloc(sizeof(char *) * MAXFILES);
	 p1 = argvec1; p2 = argvec2; p3 = argvec3;
	 for (n = 0; n < MAXFILES; n++){
	    *p1++ = NULL;
	    *p2++ = NULL;
	    *p3++ = NULL;
	    }
	 for (typ = 1; typ < NTYPES; ){
	    dload(argvec1, typ++);
	    dload(argvec2, typ++);
	    dload(argvec3, typ++);
	    dshow(argvec1, argvec2, argvec3);
	    }
      }

/* dload - load in a single argvec */
   static dload(a, t)
      char **a;
      int t;
      
      {
         int n, compar();
	 struct elem *e;
	 char **astart;
         
	 if (e_files[t] != NULL){
	    *a++ = FTYPES[t];
	    *a++ = UTYPES[t];
	    astart = a;
	    for (n = 0, e = e_files[t]; e != NULL; e = e->next){
	       *a++ = e->fname;
	       if (++n >= MAXFILES || e == e_files[t]->prev)
		  break;
	       }
	    *a = NULL;
	    qsort(astart, n, sizeof(char *), compar);
	    }
	 else
	    *a = NULL;
      }

/* dshow - show three argvecs */
   static dshow(a, b, c)
      char **a, **b, **c;
      
      {
         char **p, *buf, line[FB_MAXLINE], temp[FB_MAXLINE];
	 int output = 0;
	 
	 for (;;){
	    line[0] = temp[0] = NULL;
	    if (!(*a) && !(*b) && !(*c))
	       break;
	       
	    buf = NIL;
	    if (*a)
	       buf = *a++;
	    strcat(line, fb_pad(temp, buf, 26));
	       
	    buf = NIL;
	    if (*b)
	       buf = *b++;
	    strcat(line, fb_pad(temp, buf, 26));
	       
	    buf = NIL;
	    if (*c)
	       buf = *c++;
	    strcat(line, fb_pad(temp, buf, 26));
	    if (strlen(line) > 0){
	       puts(line);
	       output = 1;
	       }
	    }
	 if (output)
	    puts(NIL);
      }

/* compar - compar two **char objects */
   static compar(a, b)
      char **a, **b;
      
      {
         return(strcmp(*a, *b));
      }

/* afile - determine if f is a cdb file */
   static void afile(f)
      char *f;
      
      {
         char *ext;
	 int typ;
	 
	 ext = strrchr(f, '.');
	 if (ext == 0)
	    return;
	 *ext++ = NULL;
	 if (equal(ext, "cdb"))
	    typ = 1;
	 else if (equal(ext, "ddict"))
	    typ = 2;
	 else if (equal(ext, "map"))
	    typ = 3;
	 else if (equal(ext, "idx"))
	    typ = 4;
	 else if (equal(ext, "idict"))
	    typ = 5;
	 else if (equal(ext, "idicti"))
	    typ = 6;
	 else if (equal(ext, "vdict"))
	    typ = 7;
	 else if (equal(ext, "sdict"))
	    typ = 8;
	 else if (equal(ext, "idictc"))
	    typ = 9;
	 else if (equal(ext, "prt"))
	    typ = 10;
	 else if (equal(ext, "idictp"))
	    typ = 11;
	 else if (equal(ext, "emit"))
	    typ = 12;
	 else if (equal(ext, "lbl"))
	    typ = 13;
	 else if (equal(ext, "idictl"))
	    typ = 14;
	 else if (equal(ext, "idictu"))
	    typ = 15;
	 else if (equal(ext, "nodel"))
	    typ = 16;
	 else if (equal(ext, "noadd"))
	    typ = 17;
	 else
	    return;
	 install(f, typ);
      }

/* install f as a cdb file of typ */
   static void install(f, typ)
      char *f;
      int typ;
      
      {
         struct elem *e, *h;
	 
	 e = (struct elem *) fb_malloc(sizeof(struct elem));
	 e->fname = NULL;
	 fb_mkstr(&(e->fname), f);
	 if (e_files[typ] == NULL){
	    e_files[typ] = e;
	    e->next = e->prev = e;
	    return;
	    }
	 e->next = e_files[typ];
	 e->prev = e_files[typ]->prev;
	 e_files[typ]->prev->next = e;
	 e_files[typ]->prev = e;
      }
